ngraph/core/reference/include/ngraph/runtime/reference/gather_nd.hpp

```diff
@@ -26,171 +28,106 @@ namespace ngraph
+                    const auto end_s = next(beg_s, indices_shape.back() - 1);
+                    for (auto s = beg_s; s != end_s; ++s)
+                    {
+                        const auto o = offsets.back() * *s;
+                        offsets.push_back(o);
+                    }
+
+                    std::reverse(begin(offsets), end(offsets));
+                    return offsets;
+                }();
+
+                assert(indices_shape.back() == indices_offsets.size());
+
+                const auto batch_offset = indices_offsets.front() * params_shape[batch_dims];
+
+                const Shape k_1_indecies(next(begin(indices_shape), batch_dims),
+                                         prev(end(indices_shape)));
+
+                const Shape k_1_params(next(begin(params_shape), batch_dims),
+                                       prev(end(params_shape)));
+
+                const auto number_of_slices_to_copy_in_one_batch = shape_size(k_1_indecies);
+
+                const auto coordinates_size = indices_shape.back();
+
+                for (size_t b = 0; b != batch_size; ++b)
                 {
-                    if (params_coord_iter == params_outer_transform.end() ||
-                        out_coord_iter == out_transform.end())
-                        break;
-                    auto indices_index = indices_outer_transform.index(indices_coord);
-                    auto params_index = params_outer_transform.index(*params_coord_iter);
-                    auto output_index = out_transform.index(*out_coord_iter);
-                    gather_nd_batch(params + params_index,
-                                    indices + indices_index,
-                                    out + output_index,
-                                    params_shape_batch,
-                                    indices_shape_batch,
-                                    output_shape_batch);
-
-                    out_coord_iter++;
-                    params_coord_iter++;
+                    const auto input_batch_offset = b * batch_offset;
+                    const auto output_batch_offset =
+                        b * number_of_slices_to_copy_in_one_batch * slice_size;
+                    const auto coordinates_batch_offset =
+                        b * number_of_slices_to_copy_in_one_batch * coordinates_size;
+                    for (size_t i = 0; i != number_of_slices_to_copy_in_one_batch; ++i)
+                    {
+                        const auto coordinates =
+                            indices + coordinates_batch_offset + i * coordinates_size;
+
+                        size_t mem = input_batch_offset;
+                        for (size_t c = 0; c != coordinates_size; ++c)
+                        {
+                            const auto i_c = coordinates[c];
+                            const auto index = i_c < 0 ? k_1_params[c] + i_c : i_c;
+                            mem += index * indices_offsets[c];
```

Comment:
This loop looks like a std::accumulate but I'm not sure if switching to algo would affect the performance. Just a suggestion to test it though.

Reply:
I'm not sure if is `accumulate` algo which iterate over three different containers.